packages/lp32/reader.js

```diff
@@ -0,0 +1,88 @@
+// @ts-check
+/// <reference types="ses"/>
+
+// We use a DataView to give users choice over endianness.
+// But DataView does not default to host-byte-order like other typed arrays.
+
+import { hostIsLittleEndian } from './src/host-endian.js';
+
+const { details: X, quote: q } = assert;
+
+/**
+ * @param {Iterable<Uint8Array> | AsyncIterable<Uint8Array>} reader
+ * @param {Object} [opts]
+ * @param {string=} [opts.name]
+ * @param {number} [opts.maxMessageLength] - defaults to 1MB
+ * @param {boolean=} [opts.littleEndian]
+ * @param {number=} [opts.initialCapacity]
+ * @returns {import('@endo/stream').Reader<Uint8Array, undefined>}
+ */
+async function* makeLp32Iterator(
+  reader,
+  {
+    name = '<unknown>',
+    initialCapacity = 1024,
+    maxMessageLength = 1024 * 1024, // 1MB
+    littleEndian = hostIsLittleEndian,
+  } = {},
+) {
+  let capacity = Math.max(4, initialCapacity);
+  let length = 0;
+  let array8 = new Uint8Array(capacity);
+  let data = new DataView(array8.buffer);
+  let offset = 0;
+
+  for await (const chunk of reader) {
+    if (length + chunk.byteLength >= capacity) {
+      while (length + chunk.byteLength >= capacity) {
+        capacity *= 2;
+      }
+      const replacement = new Uint8Array(capacity);
+      replacement.set(array8, 0);
+      array8 = replacement;
+      data = new DataView(array8.buffer);
+    }
+    array8.set(chunk, length);
+    length += chunk.byteLength;
+
+    let drained = false;
+    while (!drained && length >= 4) {
+      const messageLength = data.getUint32(0, littleEndian);
+      assert(
+        messageLength <= maxMessageLength,
+        X`Messages on ${q(
+          name,
+        )} must not exceed ${maxMessageLength} bytes in length`,
+      );
+      const envelopeLength = 4 + messageLength;
+      if (envelopeLength <= length) {
```

Comment:
I'd prefer the following, but if it doesn't seem to make things clearer for you, feel free to disregard.

```suggestion
      drained = envelopeLength > length;
      if (!drained) {
```